mitmproxy/command.py

```diff
@@ -159,23 +163,29 @@ def parse_partial(
         """
             Parse a possibly partial command. Return a sequence of ParseResults and a sequence of remainder type help items.
         """
-        buf = io.StringIO(cmdstr)
         parts: typing.List[str] = []
-        lex = lexer(buf)
-        while 1:
-            remainder = cmdstr[buf.tell():]
-            try:
-                t = lex.get_token()
-            except ValueError:
-                parts.append(remainder)
-                break
-            if not t:
-                break
-            parts.append(t)
+
+        rex = pyparsing.QuotedString("\"", escChar='\\', unquoteResults=False) |\
```

Comment:
@BkPHcgQL3V did some previous work that showed we're on the hot path here and need to be a bit mindful of performance.

Perf 1: Move the whole parsing step into an isolated function that can be cached with `functools.lru_cache`.
Perf 2: Create `rex` once only and then re-use it.

Reply:
Hmmm... not sure why this is part of the hot path. It is supposed to be called when you are executing a command. Anyways, I separated it in a different function. How do I make it cacheable by functools.lru_cache?

Reply:
Alright, just read the documentation for lru_cache. I have my doubts about the effectiveness of it though considering that every time a letter is typed this function will be called, so if the user decides to type the expression:

`: set view_filter=~bq "(?i)[a-z]{2,3}"`

This will take 36 entries in the cache already.

There are 2 commands that I see calling this function like once a second each: `view.properties.marked` and `view.properties.length`. For these two, the cache might help a bit, but I think that is the only occasion.

Just to clarify, this is the function I created:

```python
def _parse_cmd(cmdstr: str):

    import mitmproxy.ctx as ctx

    if ctx.log:
        ctx.log.info(cmdstr)
    parts: typing.List[str] = []

    regex = pyparsing.QuotedString("\"", escChar='\\', unquoteResults=False) |\
        pyparsing.QuotedString("'", escChar='\\', unquoteResults=False) |\
        pyparsing.Combine(pyparsing.Literal('"') + pyparsing.Word(pyparsing.printables + " ") + pyparsing.StringEnd()) |\
        pyparsing.Word(pyparsing.printables) |\
        pyparsing.Word(' ')

    regex = regex.leaveWhitespace()

    for t, start, end in regex.scanString(cmdstr):
        parts.append(t[0])

    if not parts:
        parts = []

    return parts
```

Reply:
> I have my doubts about the effectiveness of it though considering that every time a letter is typed this function will be called, so if the user decides to type the expression:

You are right - it doesn't matter for users typing commands and we wouldn't need to/shouldn't cache here. It does matter for rendering the flow list: Some commands are called on each render (such as `view.properties.marked`), and this adds up when scrolling quickly or when lots of new flows are arriving.

Reply:
Second amendment: let's move the creation of `regex` out of the function definition, that object can be re-used. :)

Reply:
Done!